### PR TITLE
[JBIDE-21428] look up oc binary in preferences and fall back to $PATH

### PR DESCRIPTION
**Pull request by adietish**

### CODE REVIEW

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OCBinary.java [hunk 1]
```diff
@@ -35,20 +48,30 @@ public String getName() {
...
+	 * Returns the location from preferences or looks it up on the path.
+	 * 
+	 * @return
+	 */
+	public String getLocation() {
+		String location = OpenShiftCorePreferences.INSTANCE.getOCBinaryLocation();
+		if (StringUtils.isBlank(location)) {
+			location = findLocation();
 		}
+		return location;
```
Comment (scabanovich) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OCBinary.java]: Then OpenShiftPreferencePage and OpenShiftCorePreferenceInitializer cannot use that method, they need 'pure' search for default in $PATH.

Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OCBinary.java]: I did that, the OpenShiftCorePreferenceInitializer now calls 
```
		String location = OCBinary.getInstance().findLocation();
```
right?

**Comment by adietish** [comment]
I now also included your changes to OpenShiftPreferencePage.
@scabanovich plz review

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OpenShiftCorePreferenceInitializer.java [hunk 2]
```diff
@@ -22,9 +23,9 @@
 	@Override
 	public void initializeDefaultPreferences() {
 		IEclipsePreferences defaultPreferences = ((IScopeContext)DefaultScope.INSTANCE).getNode(OpenShiftCoreActivator.PLUGIN_ID);
-		
-		String location = OCBinaryName.getInstance().getLocation();
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OpenShiftCorePreferenceInitializer.java]: as you suggested, we're now using #findLocation (which only looks up in $PATH)

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OCBinary.java (hunk 1 above)
Comment (scabanovich) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/preferences/OCBinary.java]: Oh, yes. I would vote for getDefaultLocation(), name findLocation() remained obscure to me. But it is ok, as is done.

**Comment by scabanovich** [comment]
It looks good to me.

**Comment by adietish** [comment]
ok, thanks!

**Comment by robstryker** [comment]
What was the logic of only searching the $PATH directly? Sometimes (such as on mac) the binary may be placed in a folder that's not on the default path, or that's only on the path when using a command line shell but not eclipse's path. 

Eclipse gets a path when it is run. If the user then goes and changes the path (After discovering an error for example), do we really want him to reboot his workspace? Or do we just want to be able to find it seamlessly by running a bash command in the background to see if it's on the path now? 

The full logic of CommandLocationBinary will check the PATH passed to eclipse environment, but if it still can't find it, will run bash in the background. This allows the user to update the PATH via export PATH=etc on the command line without having to reboot eclipse.

**Pull request merged**